apparmor: try again to fix and improve

ju1m opened the pull request:
This PR is **ready to be merged**. It re-introduces work done in https://github.com/NixOS/nixpkgs/pull/93457, hopefully without the [import-from-derivation](https://nixos.wiki/wiki/Import_From_Derivation) which was: 
1. causing `hydra-eval-jobs` to crash because of https://github.com/NixOS/hydra/pull/825
2. and requiring the option `allow-import-from-derivation` to be set.

<details>
<summary>Old and wrong analysis believing <code>apparmorRulesFromClosure</code> is an IFD</summary>

Please, do not merge this PR before https://github.com/NixOS/hydra/pull/825 has been merged. It maintains a mergeable version of PR #93457, previously reverted due to a ~(so far not understood)~ problem raised by the hydra evaluation.
~It may or may not be caused by `allow-import-from-derivation = false`. More investigation is needed.~
This problem is due to the following:
1. To simplify the generation of AppArmor profiles, this PR introduces and uses `pkgs.apparmorRulesFromClosure`, which leverages IFD (import-from-derivation) via `pkgs.closureInfo`.
2. However, requiring IFD means that `hydra-eval-jobs` now [needs to perform builds](https://github.com/Gabriel439/haskell-nix/issues/6#issuecomment-270164683) simply to do its task of listing Hydra jobs in a big JSON object for `hydra-eval-jobset`. For that reason or another, @vcunat [informed us](https://github.com/NixOS/nixpkgs/pull/93457#issuecomment-705400954) that: "_import-from-derivation is better avoided in the official nixpkgs repo (not sure if there's some rule on that) and it's intentionally disabled on Hydra.nixos.org_", causing all NixOS tests depending on packages or services using `pkgs.apparmorRulesFromClosure` to fail on [hydra.nixos.org](hydra.nixos.org).
3. [hydra-eval-jobs](https://github.com/NixOS/hydra/blob/master/src/hydra-eval-jobs/hydra-eval-jobs.cc) is used on [hydra.nixos.org](hydra.nixos.org) to eval [nixos/release-combined.nix](https://github.com/NixOS/nixpkgs/blob/master/nixos/release-combined.nix) where an [aggregate optimization](https://github.com/NixOS/hydra/issues/715) is used, but if one of the `constituents` of this aggregation fails then the overall jobset generation (in JSON) fails with:
```
$ NIX_PATH= hydra-eval-jobs -I . nixos/release-combined.nix --option allow-import-from-derivation false --verbose
[…]
error: [json.exception.type_error.302] type must be string, but is null
```

The **good news** is that this bug is being fixed by @samueldr in https://github.com/NixOS/hydra/pull/825, such that `hydra-eval-jobs` now ignores failing aggregated jobs instead of crashing:
```
$ NIX_PATH= hydra-eval-jobs -I . nixos/release-combined.nix --option allow-import-from-derivation false --verbose
[…]
  "nixos.tests.transmission.x86_64-linux": {                                                                                                                                                                                                                   
    "error": "error: --- EvalError --- hydra-eval-jobs\nattempted to realize '/nix/store/f46p5hlwks57rmxdpm0jkixlq1mf5dw4-apparmor-utils-2.13.5.drv' during evaluation but 'allow-import-from-derivation' is false"                                            
  },
[…]
```
In that example I've added `nixos.tests.transmission` amongst aggregated jobs in [nixos/release-combined.nix](https://github.com/NixOS/nixpkgs/blob/master/nixos/release-combined.nix), commenting all others in [nixos/release-combined.nix](https://github.com/NixOS/nixpkgs/blob/master/nixos/release.nix) to get a quick run (because running on all jobs takes… ~5 hours on my computer). By lack of time and computer power, I was unable to bisect which aggregated tests were failing, but I was able to reproduce that crashing with all tests enabled, and no more crashing with https://github.com/NixOS/hydra/pull/825.

The **bad news** is that adding AppArmor profiles using `pkgs.apparmorRulesFromClosure` will potentially disable tests on [hydra.nixos.org](hydra.nixos.org) down the road, maybe a lot of them eventually.
If those cases are considered important enough to not be disabled, one could:
1. Use another Hydra instance where IFD is enabled, and maybe write a `nixos/release-ifd.nix` aggregating only those important tests.
2. Not use `pkgs.apparmorRulesFromClosure` on the concerned services/packages, by listing manually all the dependencies, though this could be very hard to maintain for many services/packages.
3. Write less secure AppArmor rules globbing over the whole `/nix/store`, instead of limiting them to the packages of a closure.
</details>

###### Motivation for this change
See PR #93457.

###### Things done
- [X] Revert commit 420f89ceb267b461eed5d025b6c3c0e57703cc5c
- [X] Fix `logprof.conf` generation by removing the [import-from-derivation](https://nixos.wiki/wiki/Import_From_Derivation) (IFD) caused by `builtins.readFile` on the `apparmor-utils` derivation.
- [X] Update to latest major AppArmor release : [3.0.1](https://gitlab.com/apparmor/apparmor/-/wikis/Release_Notes_3.0).
- [X] Add an optional `name` attribute to `apparmorRulesFromClosure` to make listing `/nix/store/` more human readable, so that there is some context rather than hundreds of paths designated "apparmor-closure-rules".
- [X] Fix `aa-unconfined` not finding `netstat` nor `ss` at runtime because it overwrites `PATH`.
- [X] Move `bin.transmission-daemon` profile to Nixpkgs, and customize it in NixOS with `local/bin.transmission-daemon`.
- [X] Add `nixos/tests/apparmor.nix`, testing AppArmor is enabled in the kernel and `apparmorRulesFromClosure` works.
- [X] Disable `security.apparmor.killUnconfinedConfinables` by default, but enable it in `nixos/modules/profiles/hardened.nix`.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

ju1m committed:
Revert "Revert "apparmor: fix and improve the service""

This reverts commit 420f89ceb267b461eed5d025b6c3c0e57703cc5c.

bqv commented:
Bump?

ju1m commented:
Thanks @bqv, I've rebased to merge with `iputils` changes in 0a55c5dada75480342ee9b58932d612438793302

bqv commented:
@ju1m new conflicts

Also, is there any chance you'll be able to resolve the hydra issue soon?

ju1m commented:
Rebased after `{utillinux => util-linux}` rename in https://github.com/NixOS/nixpkgs/pull/104776 .

ju1m commented:
> Also, is there any chance you'll be able to resolve the hydra issue soon?

Good news everyone! Turns out @samueldr has been tackling this `hydra-eval-jobs` issue in https://github.com/NixOS/hydra/pull/825. I've updated the intro with all the details.

bqv commented:
I imagine some semblance of option 1 is the most reasonable and realistic

bqv commented on pkgs/os-specific/linux/apparmor/default.nix:
```diff
@@ -249,8 +274,32 @@ let
+        # eg. glibc-2.30/lib/gconv/gconv-modules
+        "r $path/lib/**"
+      ]
+    }: rootPaths: runCommand "apparmor-closure-rules" {} ''
```
I think these should be named by argument, for clarity in the store

ju1m replied:
Do you mean adding something in "apparmor-closure-rules"? what would be that something?

bqv replied:
Yes, I mean pass in a designation as an argument to apparmorRulesFromClosure, so that there is some context rather than hundreds of paths designated "apparmor-closure-rules"

ju1m replied:
I've added an optional `name` attribute.

bqv commented on pkgs/os-specific/linux/iputils/default.nix:
```diff
@@ -44,6 +47,25 @@ in stdenv.mkDerivation rec {
+      include <abstractions/consoles>
+      include <abstractions/nameservice>
+      include "${apparmorRulesFromClosure {}
+       ([libcap] ++ optional (!stdenv.hostPlatform.isMusl) libidn2)}"
```
This could just be buildInputs, to dedup

ju1m replied:
Indeed, but I don't think this a good habit to take and example to provide, because the dependencies of a package are not necessarily the same as the dependencies of a specific executable of this package. For instance here the `tftpd` in the same package does not depend on `libpcap` nor `libidn2` (according to `ldd`). Same could apply to `ping` in the future if another executable needs a `buildInputs` item not required by `ping`. This mismatch is [the main reason](https://github.com/NixOS/nixpkgs/pull/93457#issuecomment-705606683) why this PR moves AppArmor profiles from NixOS to Nixpkgs.

bqv commented on pkgs/os-specific/linux/apparmor/default.nix:
```diff
@@ -140,6 +153,15 @@ let
       # aa-notify checks its name and does not work named ".aa-notify-wrapped"
       mv $out/bin/aa-notify $out/bin/aa-notify-wrapped
       makeWrapper ${perl}/bin/perl $out/bin/aa-notify --set PERL5LIB ${libapparmor}/${perl.libPrefix} --add-flags $out/bin/aa-notify-wrapped
+
```
aa-unconfined needs netstat at runtime

ju1m replied:
Indeed `aa-unconfined` can't find `netstat` nor `ss` because it completely overwrites `PATH` :facepalm: https://gitlab.com/apparmor/apparmor/-/blob/master/utils/aa-unconfined#L89

ju1m committed:
apparmor: 2.13.5 -> 3.0.0

ju1m commented:
I've added as a separate commit the updating of AppArmor to its latest major release: [3.0.0](https://gitlab.com/apparmor/apparmor/-/wikis/Release_Notes_3.0). Testers can only cherry-pick the other commit to stay on `2.13.5`.

bqv commented:
That now makes this 5k+ again. It's gonna be a full century before this is merged...

edolstra commented:
We cannot use IFD in NixOS/Nixpkgs. It's disabled on hydra.nixos.org to enforce this, not just because it's bad for Hydra but because it's bad for users (e.g. just evaluating a configuration could trigger an arbitrary amount of building). 

The workaround is to use closure info only at build time, not at evaluation time. For example, something like
```
security.apparmor.includes."nixos/security.wrappers" =
  runCommand "..." { }
    ''
      echo -n "include " > $out
      cat ${pkgs.apparmorRulesFromClosure { name="security.wrappers"; } [ 
        securityWrapper
        pkgs.stdenv.cc.cc
        pkgs.stdenv.cc.libc
      ]} >> $out
    '';
```
should work.

edolstra commented:
By the way, *where* is this PR using IFD? I'm not seeing it...

ju1m commented:
> By the way, _where_ is this PR using IFD? I'm not seeing it...

@edolstra you're on it, it is now this very one, in its first commit reverting the revert done by https://github.com/NixOS/nixpkgs/commit/420f89ceb267b461eed5d025b6c3c0e57703cc5c.

`apparmorRulesFromClosure` is defined in [pkgs/os-specific/linux/apparmor/default.nix](https://github.com/NixOS/nixpkgs/pull/101071/commits/4f311528958b9296fa303c501d21ad961a0444c2#diff-ebea69c8f7f6a46da1809b617d4214ebfaccd6ff9b30ab0cbfb444105973ba59R278-R305).
And currently used in Nixpkgs to generate a new `apparmor` output for [inetutils](https://github.com/NixOS/nixpkgs/pull/101071/commits/4f311528958b9296fa303c501d21ad961a0444c2#diff-729ed11f4fe851cb6ef34b1783b6b6027114311f910723ce5d4006930a4afb03R48-R63) and [iputils](https://github.com/NixOS/nixpkgs/pull/101071/commits/4f311528958b9296fa303c501d21ad961a0444c2#diff-ae4961273d9cde86be850ebc01f88069e8ff078b60c79f5590f8e7fe655217d3R50-R68) where only a subset of the `buildInputs` of those packages are given to `closureInfo`.

In [nixos/modules/security/apparmor/profiles.nix](https://github.com/NixOS/nixpkgs/pull/101071/commits/4f311528958b9296fa303c501d21ad961a0444c2#diff-10332acc1b0ccbb38f1d647f747487a4e7eb3e70ac408ca39634e2c9825d8213) those `apparmor` outputs are included iff `security.apparmor` and the `bin.ping` profile are enabled:
```
config.security.apparmor.policies."bin.ping".profile = lib.mkIf apparmor.policies."bin.ping".enable ''
  include "${pkgs.iputils.apparmor}/bin.ping"
  include "${pkgs.inetutils.apparmor}/bin.ping"
'';
}
```

`apparmorRulesFromClosure` is also directly used in NixOS for:
- [security.apparmor.policies."bin.transmission-daemon".profile](https://github.com/NixOS/nixpkgs/pull/101071/commits/4f311528958b9296fa303c501d21ad961a0444c2#diff-4c85e2362ce5f3853f62aa5b8b0d3ab9b46e449505274cfffed671124e94edaeR367-R369) but it would be [more correct](https://github.com/NixOS/nixpkgs/pull/93457#issuecomment-705606683) to be in `pkgs.transmission-daemon.apparmor`, and I should change that.
- [security.apparmor.includes."nixos/security.wrappers"](https://github.com/NixOS/nixpkgs/pull/101071/commits/4f311528958b9296fa303c501d21ad961a0444c2#diff-34fec612f75768c480ca1cbc6979c70710df345a951713c9291894ae8fa17b07R182-R189), because `security.wrappers` only makes sense in NixOS.

bqv commented:
> We cannot use IFD in NixOS/Nixpkgs

@edolstra AIUI, closureInfo is already used in NixOS, and I think I'm right in thinking that that's the source of the IFD in this PR? Is it then a rule that closureInfo *cannot* be used in nixpkgs (and if so why does it exist?)

bqv commented on nixos/modules/security/apparmor.nix:
```diff
@@ -1,59 +1,201 @@
+        ${pkgs.bashInteractive}/bin/sh = icnu
+        ${pkgs.bashInteractive}/bin/bash = icnu
+    '' + head (match "^.*\\[qualifiers](.*)" # Drop the original [settings] section.
+                     (readFile "${pkgs.apparmor-utils}/etc/apparmor/logprof.conf"));
```
Or is this IFD?

bqv replied:
Refactored in ju1m/nixpkgs#12. If that's the issue I'm lowlevel pissed nobody cared to notice earlier.

ju1m replied:
Well found @bqv! This is indeed the IFD, it was not `closureInfo` as [previously thought](https://github.com/NixOS/nixpkgs/pull/93457#issuecomment-704920361). I had understood @edolstra's sentence as asking where was the PR, but he was in fact asking where was the IFD. 
I was able to make `hydra-eval-jobs` pass `nixos.tests.hardened` and `nixos.tests.transmission` with https://github.com/ju1m/nixpkgs/pull/12 , and make it crash without.

bqv committed:
nixos/apparmor: remove an IFD

First because IFD (import-from-derivation) is not allowed on hydra.nixos.org,
and second because without https://github.com/NixOS/hydra/pull/825
hydra-eval-jobs crashes instead of skipping aggregated jobs which fail
(here because they required an IFD).

bqv commented:
There is no longer anything blocking this PR.

@vcunat as the one who reverted the previous iteration of this, do you have any reason not to merge this now?

ju1m committed:
nixos/apparmor: fix logprof.conf generation

ju1m committed:
transmission: move apparmor profile to Nixpkgs

bqv commented:
Primarily @vcunat, but also
@edolstra 
@thoughtpolice
@matthewbauer
@joachifm
@Phreedom
@lheckemann
@primeos

Would one of you please describe what more this PR needs to do to be done?

vcunat commented:
Primarily me?  I believe my only connection is that I reverted the previous version in commit 420f89c.  That was an emergency to unblock the build farm; I didn't _really_ understand the contents of the reverted work, and I know basically nothing about apparmor... so it's not easy for me to review this PR.

_I'm generally sorry for any bigger piece of work that remains unmerged for long, but honestly I don't feel motivated to dig deep here in comparison to other nixpkgs PRs._

ju1m commented:
@7c6f434c as you merged this PR once, could you consider remerging it now that it should comply with hydra.nixos.org ? thanks in advance

bqv commented:
@ju1m it's been so long again that apparmor-parser needs a patch. See https://github.com/bqv/apparmor-nix/commit/34b98f6bd70c37e68cec9a2d7e33d3aa7278429b (orig https://github.com/vmware/photon/commit/f9b0c595)

7c6f434c commented:
I basically merged it because it was clear that nobody cared enough about this, but now I have no idea how to check Hydra-safety of this…

ju1m commented:
> I basically merged it because it was clear that nobody cared enough about this, but now I have no idea how to check Hydra-safety of this…

This is the command that was used to check Hydra compliance in the intro of this PR; it may take several hours depending on your computer, mainly because it evaluates a lot of Nix expressions.
```
hydra-eval-jobs -I . nixos/release-combined.nix --option allow-import-from-derivation false --verbose
```

Flakebi commented:
I tried `hydra-eval-jobs -I . nixos/release-combined.nix --option allow-import-from-derivation false --verbose` and it finished successfully after 4.5 hours.
Building a system fails as @bqv pointed out as `apparmor-parser-3.0.0` fails to build.

ju1m committed:
apparmor: 3.0.0 -> 3.0.1

ju1m commented:
> I tried `hydra-eval-jobs -I . nixos/release-combined.nix --option allow-import-from-derivation false --verbose` and it finished successfully after 4.5 hours.
> Building a system fails as @bqv pointed out as `apparmor-parser-3.0.0` fails to build.

@Flakebi thanks for the report, I've added a commit updating AppArmor from `3.0.0` to `3.0.1` which [should fix](https://gitlab.com/apparmor/apparmor/-/wikis/Release_Notes_3.0.1#highlighted-fixes-or-improvements) the build on Linux `>= 5.9`.

Flakebi commented:
Thanks, I tried it with `build-vm` and it works good (also tried with a gui). I don’t use apparmor much, so only the ping profile was loaded.
As this rebuilds quite a few things, I guess the pr should go into `staging` instead of directly into `master`.

Flakebi commented on pkgs/os-specific/linux/apparmor/default.nix:
```diff
@@ -26,10 +33,16 @@ let
   };
 
   apparmor-sources = fetchurl {
-    url = "https://launchpad.net/apparmor/${apparmor-series}/${apparmor-version}/+download/apparmor-${apparmor-version}.tar.gz";
-    sha256 = "13xshy7905d9q9n8d8i0jmdi9m36wr525g4wlsp8k21n7yvvh9j4";
+    url = "https://launchpad.net/apparmor/${apparmor-series}/${apparmor-series}.${apparmor-patchver}/+download/apparmor-${apparmor-version}.tar.gz";
+    sha256 = "096zbg3v7b51x7f1ly61mzd3iy9alad6sd4lam98j2d6v5ragbcg";
```
Nitpick: Maybe the most idiomatic way here is to have `apparmor-version = "3.0.1"` instead of `-series` and `-patchver` and use `${lib.versions.majorMinor apparmor-version}` instead of `${apparmor-series}`?

ju1m replied:
Thanks

Flakebi reviewed (commented):
Can you change the base branch of the pull request to `staging` instead of `master`?
I think mass rebuilds should go there (that’s what I read from this nice diagram: https://github.com/NixOS/rfcs/pull/26#issuecomment-739237393).

ju1m commented:
Rebased against `staging` as pointed out by @Flakebi

ju1m committed:
nixos/apparmor: move release note to 21.05

Flakebi commented:
Thanks!
I think you can also change the branch in the pull request. There should be an edit button on the top right. If you press that one, the base branch becomes a drop-down where a branch can be selected.

Flakebi reviewed (approved):
Looks good to me.
I noticed that the apparmor output files of ping are marked as executable, I guess this is just the default behaviour of `install -D` and not on purpose (the mode can be overwritten with `-m 644`).

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/328

ju1m commented:
- Fixed `install -D` as pointed out by @Flakebi.
- Re-added forgotten commit about `transmission`.

ju1m commented:
Rebased against `staging` after merge conflict in release notes.
@SuperSandro2000 sorry for the delay, I do not receive mail when a label is added.

SuperSandro2000 commented:
> @SuperSandro2000 sorry for the delay, I do not receive mail when a label is added.

I added the automatically to several hundred PRs.

SuperSandro2000 commented on pkgs/os-specific/linux/apparmor/default.nix:
```diff
@@ -121,7 +126,11 @@ let
-    prePatch = prePatchCommon + ''
+    prePatch = prePatchCommon +
+      # Do not build vim file
+      stdenv.lib.optionalString stdenv.hostPlatform.isMusl ''
```
```suggestion
      lib.optionalString stdenv.hostPlatform.isMusl ''
```

ju1m replied:
Thanks, I've included those suggestions.

SuperSandro2000 commented on pkgs/os-specific/linux/apparmor/default.nix:
```diff
@@ -249,8 +269,35 @@ let
+    , name ? ""
+    }: rootPaths: runCommand
+      ( "apparmor-closure-rules"
+      + stdenv.lib.optionalString (name != "") "-${name}") {} ''
```
```suggestion
      + lib.optionalString (name != "") "-${name}") {} ''
```

ju1m commented:
Rebased after 25e3e66b652f39cadad08f2f2aadbe7b2f8211b7's changes in `pkgs/os-specific/linux/iputils/default.nix`

Flakebi commented on pkgs/os-specific/linux/iputils/default.nix:
```diff
@@ -34,6 +37,25 @@ in stdenv.mkDerivation rec {
+      include <abstractions/consoles>
+      include <abstractions/nameservice>
+      include "${apparmorRulesFromClosure { name = "ping"; }
+       ([libcap] ++ optional (!stdenv.hostPlatform.isMusl) libidn2)}"
```
ofborg complains that `optional` is not defined here. Probably should be `lib.optional`

Flakebi commented:
Tested in a vm and still works great,
@Mic92 can you merge this please?

Mic92 commented:
> Tested in a vm and still works great,
> @Mic92 can you merge this please?

Having `killUnconfinedConfinables` sounds still scary to me. What if it kills your network daemon/vpn while upgrading apparmor rules and you no longer can reach the same machine.

andir commented on nixos/modules/security/apparmor.nix:
```diff
@@ -1,59 +1,207 @@
-     };
-   };
+  imports = [
+    (lib.mkRenamedOptionModule [ "security" "virtualization" "flushL1DataCache" ] [ "security" "virtualisation" "flushL1DataCache" ])
```
Maybe this rename should be moved to where the new option is defined after all? It looked misplaced before and still does.

ju1m replied:
indeed, thanks

andir commented on nixos/modules/security/apparmor.nix:
```diff
@@ -1,59 +1,207 @@
+
+  options = {
+    security.apparmor = {
+      enable = lib.mkEnableOption ''the AppArmor Mandatory Access Control system.
```
Can you please add a newline before the text? Nix should strip out everything before the first common indent character.

ju1m replied:
oh right, it is required when `"` is used but not with `''`

andir commented on nixos/modules/security/apparmor.nix:
```diff
@@ -1,59 +1,207 @@
+        default = [];
+        description = "List of packages to be added to AppArmor's include path";
+      };
+      enableCache = lib.mkEnableOption ''caching of AppArmor policies
```
Same as above. Please add a newline here to make this more eye friendly.

andir commented on nixos/modules/security/apparmor.nix:
```diff
@@ -1,59 +1,207 @@
+        Beware that AppArmor policies almost always contain Nix store paths,
+        and thus produce at each change of these paths
+        a new cached version accumulating in the cache'';
+      killUnconfinedConfinables = mkDisableOption ''killing of processes
```
same

andir commented on nixos/modules/security/apparmor.nix:
```diff
@@ -1,59 +1,207 @@
+    environment.etc."apparmor.d".source = pkgs.linkFarm "apparmor.d" (
+      # It's important to put only enabledPolicies here and not all cfg.policies
+      # because aa-remove-unknown reads profiles from all /etc/apparmor.d/*
+      lib.mapAttrsToList (name: p: {inherit name; path=p.profile;}) enabledPolicies ++
```
This  could do with some reformatting. It looks pretty cramped. At first it looked like you tried to check for equality but forgot an equals sign.

ju1m replied:
I don't understand the reformat you're suggesting here. I'm adding spaces around curly braces and the equal sign.

andir commented on nixos/modules/security/apparmor.nix:
```diff
@@ -1,59 +1,207 @@
+      Include /etc/apparmor.d
+    '' +
+    lib.concatMapStrings (p: "Include ${p}/etc/apparmor.d\n") cfg.packages;
+    # For aa-logprof
```
Not sure why that is for that. Mind explaining that in the comment before you forget? :-)

ju1m replied:
Sure, but what don't you understand precisely?
The `/etc/apparmor.d` created a few lines above is here included, along with the `/etc/apparmor.d` of some packages as documented in the option.
I'll indent the string before the `+` and the `concatMapStrings` to group them better.

andir commented on nixos/modules/security/apparmor.nix:
```diff
@@ -1,59 +1,207 @@
+    # For aa-logprof
+    environment.etc."apparmor/apparmor.conf".text = ''
+    '';
+    # For aa-logprof
```
Same

ju1m replied:
I don't understand what you'd like here. The comment says `apparmor/apparmor.conf` is for `aa-logprof`, what's more to say? Running `aa-logprof` requires this file.

andir commented on nixos/modules/security/apparmor.nix:
```diff
@@ -1,59 +1,207 @@
+          # 3 - force all perms on the rule to be user
+          default_owner_prompt = 1
+
+          custom_includes = /etc/apparmor.d ${lib.concatMapStringsSep " " (p: "${p}/etc/apparmor.d") cfg.packages}
```
Use `lib.makeSearchPath "etc/appamrmor.d" cfg.packages` here.

ju1m replied:
That would not be correct because `makeSearchPath` uses `concatStringsSep ":"` not `concatStringsSep " "`.

andir commented on nixos/modules/security/apparmor.nix:
```diff
@@ -1,59 +1,207 @@
+          custom_includes = /etc/apparmor.d ${lib.concatMapStringsSep " " (p: "${p}/etc/apparmor.d") cfg.packages}
+
+        [qualifiers]
+          ${pkgs.runtimeShell} = icnu
```
You probably want to cover `config.users.defaultUserShell` as well here?

ju1m replied:
Indeed, thanks.

andir commented on nixos/modules/security/apparmor/includes.nix:
```diff
@@ -0,0 +1,301 @@
+  inherit (builtins) attrNames hasAttr isAttrs;
+  inherit (lib) getLib;
+  inherit (config.environment) etc;
+  etcRule = arg:
```
I can read what this is doing but do you mind adding a comment here? You probably wrote this function as it is "abstracting" some complexity away from the code further down this file. Now I just have to grok what that complexity is.

ju1m replied:
Sure.

andir commented on nixos/modules/security/apparmor/includes.nix:
```diff
@@ -0,0 +1,301 @@
+    ${etcRule "asound.conf"}
+    ${etcRule "esound/esd.conf"}
+    ${etcRule "libao.conf"}
+    ${etcRule {path="pulse"; trail="/";}}
```
Again the common style here is to add spaces around the equals signs. Would serve readability.

ju1m replied:
Done.

andir commented on nixos/modules/security/apparmor/includes.nix:
```diff
@@ -0,0 +1,301 @@
+    include "${pkgs.apparmor-profiles}/etc/apparmor.d/abstractions/fonts"
+    ${etcRule {path="fonts"; trail="/**";}}
+  '';
+  "abstractions/gnome" = ''
```
Perhaps this would look better as a nix list of attribute sets that are then translated once with once function instead of doing string interpolation over and over again?

ju1m replied:
Indeed `include` and `etcRule` calls could be factorized by generating from a list of various attributes like `includeUpstream`, `textBefore`, `etcRules`, `textAfter`, etc., but what would be the benefit of rigidifyng this code? and what is the problem with repeated string interpolations?

ju1m replied:
Or maybe I can just factorize the places where there are many `etcRule`s, I'll try and you'll tell me.

andir commented on nixos/modules/security/apparmor/profiles.nix:
```diff
@@ -0,0 +1,11 @@
+{
+config.security.apparmor.packages = [ pkgs.apparmor-profiles ];
+config.security.apparmor.policies."bin.ping".profile = lib.mkIf apparmor.policies."bin.ping".enable ''
+  include "${pkgs.iputils.apparmor}/bin.ping"
```
Does this work with our ping that lives in `/run/wrappers`?

ju1m replied:
Yes, but the relevant code is where `ping`'s wrappers are added, in `nixos/modules/tasks/network-interfaces.nix`: https://github.com/NixOS/nixpkgs/pull/101071/files#diff-0880fae3bbdeb321ece213e5fdb68064fc7eade20cdcfd27b14268e811e36d77

andir commented on nixos/modules/security/pam.nix:
```diff
@@ -895,6 +895,61 @@ in
+    security.apparmor.includes."abstractions/pam" = let
+      isEnabled = test: fold or false (map test (attrValues config.security.pam.services));
+      in ''
+      ${lib.concatMapStringsSep "\n"
```
We have to do something about readability again. I only see ${}${{}$} here and can't really figure out what might be right.

ju1m replied:
I've rewritten using `''` and `+`. Concerning the rightness, this is allowing PAM to load the PAM modules enabled in the Nix configuration.

andir commented on pkgs/os-specific/linux/apparmor/default.nix:
```diff
@@ -249,8 +269,35 @@ let
+  # from the closure of given rootPaths.
+  # To be included in an AppArmor profile like so:
+  # include "$(apparmorRulesFromClosure {} [pkgs.hello]}"
+  apparmorRulesFromClosure =
```
We should have a test for this if we offer it as an API to create closures. Doesn't have to be huge but should test that
* We do not include everything
* We do include what was passed in (for etc, share, lib, ...)

ju1m replied:
I've added a `nixos/tests/apparmor.nix` testing `apparmorRulesFromClosure`, by:
1. prepending a string of the relevant packages' name and version on each line,
2. sorting the lines according to these prepended strings,
3. removing the prepended strings,
4. using `diff` against the expected output.

andir reviewed (commented):
This is huge and i am sure was a lot of work. Thank you for working on this this is something we should have but nobody has taken on the load of work yet.

I left a few comments as I feel it is hard to read the code as someone that hasn't written it.

We *must* have a test that verifies that a basic nixos system is still bootable with apparmor enabled and that all the kill logic works as expected.

8573 commented on nixos/doc/manual/release-notes/rl-2105.xml:
```diff
@@ -662,6 +662,24 @@ self: super:
+     and to select whether to confine in enforce mode (default)
+     or in complain mode (see <literal>journalctl -b --grep apparmor</literal>).
+     Before enabling this module, either directly
+     or by importing <literal>&lt;nixpkgs/nixos/modules/profiles/hardened.nix&gt;</literal>,
```
Thank you for continuing this work. However, I think I didn't realize this last year — it's not merely people knowingly enabling AppArmor for the first time but all users of the hardened profile¹ who, upon updating (perhaps automatically) to a version of nixpkgs that includes these changes, will get any already-running confinable processes killed? What's the risk of data loss, especially with `system.autoUpgrade`?

<small>¹ unless they've specifically disabled AppArmor for some reason</small>

ju1m replied:
@8573, yes, the `hardened` profile is enabling `AppArmor`, hence people importing it would have their unconfined and newly confinable processes killed by default. Because `AppArmor` cannot confine them without killing them, enabling `killUnconfinedConfinables` by default is indeed privileging security over stability. I would argue that it respects the expectation of users importing a profile named: `hardened`. It would be more problematic if `AppArmor` were to be enabled by default on NixOS. So I guess the sensible thing is to disable `killUnconfinedConfinables` by default but enable it in the `hardened` profile, so that it can be enabled conciously by users privileging security, while privileging stability for all other users in case `AppArmor` becomes enabled by default one day.

Concerning data loss, `killUnconfinedConfinables` uses `kill(1)` to send a `SIGTERM`, not a `SIGKILL`, hence the processes have a chance to properly terminate without data loss or corruption, but not all are coded to graciously terminate like so. Ideally, they should also be restarted by `systemd`, but the service can be configured otherwise, and this is only for processes supervised by `systemd` or an other mecanism.

Concerning `system.autoUpgrade`, the introduction of a new (enabled by default) `AppArmor` profile in NixOS will indeed try to kill the processes matching that profile.

ju1m committed:
nixos/apparmor: improve code readability

ju1m committed:
nixos/apparmor: add test for apparmorRulesFromClosure

ju1m committed:
nixos/apparmor: disable killUnconfinedConfinables by default

ju1m commented:
Rebased after merge conflict in `nixos/modules/virtualisation/lxd.nix` due to changes in 4adcb0064200facdee1109a0296905717d046b3e

ju1m commented:
Well, is there anything remaining to be discussed, done or tested to get this PR merged? Please tell me if I've missed something.
@andir, have I addressed your concerns well enough?

mkg20001 commented:
I'm not sure, can we just merge this already? I'd be in favor of just merging this once conflicts are resolved, as I don't see any objections. I'd do that once conflicts are fixed.

necessarily-equal reviewed (changes requested):
Fancy! Note I didn't take time to test this yet

ju1m commented on nixos/modules/security/apparmor.nix:
```diff
@@ -1,59 +1,214 @@
+        introducing an AppArmor profile for the executable of a running process.
+
+        Enable <xref linkend="opt-security.apparmor.killUnconfinedConfinables"/>
+        if you want this service to send a </literal>SIGTERM</literal> to those running processes'';
```
Thanks!

ju1m commented on nixos/modules/profiles/hardened.nix:
```diff
@@ -36,6 +36,7 @@ with lib;
   security.virtualisation.flushL1DataCache = mkDefault "always";
 
   security.apparmor.enable = mkDefault true;
+  security.apparmor.killUnconfinedConfinables = mkDefault true;
```
Well, it does not appear to contradict the doc to me, though I've tried to clarify a bit, but tell me if I've missed somthing or if that's not clear enough. See also https://github.com/NixOS/nixpkgs/pull/101071#discussion_r584374591 where I explain why I proposed to still set `killUnconfinedConfinables` in the `hardened.nix` profile.

ju1m commented:
I've rebased against latest `staging` branch which only had minor conflicts.

7c6f434c commented:
@ofborg eval

7c6f434c commented:
@ofborg eval

aaronjanse commented:
Thank you @ju1m! I'm very excited for this to be merged

7c6f434c commented:
@ofborg test apparmor

Flakebi commented:
It is part of the current `staging-next` pr #122068, which should be merged before the 21.05 branch-off. So, unless anything unexpected happens, it will be part of 21.05. [Release schedule](https://discourse.nixos.org/t/21-05-release-schedule/12528)

vojta001 commented:
Ahhhh, great! Thanks for clarification. I am still learning the nixpkgs release process.

vcunat commented:
This PR uses `install` with `/dev/stdin` on a few places, and apparently that [will break](https://nix-cache.s3.amazonaws.com/log/gysqjdkqkxkck753hlcy85vhgw93gagg-transmission-3.00.drv) build on darwin:
```
install: skipping file '/dev/stdin', as it was replaced while being copied
```
BTW, there's no point in setting permissions there (except for the executable bit), as all nix store contents will be normalized anyway.

ju1m commented:
@vcunat see https://github.com/NixOS/nixpkgs/pull/123005 for a fix.
